### ISSUE TITLE
OAuth 2.0 - Client credentials always returning "'undefined' is not valid JSON"

### ISSUE THREAD

**Issue by eduardoscamargo** [usebruno/bruno#2452]
### I have checked the following:

- [X] I use the newest version of bruno.
- [X] I've searched existing issues and found nothing related to my issue.

### Describe the bug

I am trying to send a request using OAuth2 - Client Credentials. No matter the content of the credentials parameters I get a message `Error invoking remote method 'send-http-request': SyntaxError: "undefined" is not valid JSON`.

(Screenshot has mocked values, I tested with the real/correct values and got the same error message)

Version: v1.19.0

### Screenshots/Live demo link

<img width="1481" alt="image" src="https://github.com/usebruno/bruno/assets/8027905/6e30ff85-ad28-44f8-9e3d-fb3c24678eab">

**Comment by gemiusz** [comment]
@eduardoscamargo  I can't reproduce this:
![image](https://github.com/usebruno/bruno/assets/30750006/90e55803-9e00-4bf0-b5a4-af9280d6c5d2)



Please provide bru file with this issue. 

I'm not sure how this functionality should work so need explanation from any DEV :)
As I see when I set URL as 'test' then is going POST to 127.0.0.1:80 :
```
2024-06-14 18:50:19.391 Request received:
127.0.0.1 - POST /test

Accept: [application/json, text/plain, */*]
Content-Type: [application/x-www-form-urlencoded]
request-start-time: [1718383819286]
User-Agent: [axios/1.7.2]
Content-Length: [74]
Accept-Encoding: [gzip, compress, deflate, br]
Host: [localhost]
Connection: [close]
grant_type=client_credentials&client_id=test&client_secret=test&scope=test


```

So I think that in your case something send response that cannot be parsed correctly.
Can you provide this response ?

**Comment by sylvainlegault** [comment]
Not sure if that can help but I have the same issue, this seems to happen when using GraphQL as a simple post works for the same oauth2 settings.
There is nothing logged in the console nor in Timeline just showing < Error
Looks like the request are never sent. Might be because Graphql is not JSON

**Comment by sylvainlegault** [comment]
Did some troubleshooting, but I don't know the code base enough :(

But the problem seems to be related to this part from ipc/network/index.js
```

    // if this is a graphql request, parse the variables, only after interpolation
    // https://github.com/usebruno/bruno/issues/884
    if (request.mode === 'graphql') {
      request.data.variables = JSON.parse(request.data.variables);
    }
```
I tried skipping this but doesn't solve the issue, I get further without exception but I get the auth response instead of the graphql response, so something else is not working. Looks like oauth2 doesn't play well with graphql and doesn't continue with the grapqh request itself.

I cannot share my environment since this is a private endpoint. But I can help with testing any fixes if needed.

Thanks

**Comment by pitazzo** [comment]
any ETA on this? this is our main stopper on adopting Bruno company-wide

**Comment by VladBozhenok** [comment]
The workaround that works for me is to use Bearer Token authentication, looks silly, but it works in my case

Preconditions:
I need to use GraphQL with OAuth 2.0 Password Credentials flow

1. Create GraphQL request in Bruno
2. Generate access token in some other client (postman, insomnia, curl etc)
3. Set Bearer Token as authentication method in Bruno
4. Paste generated token to Bruno
5. Run qurequestry in Bruno

And it works, but you are dependent on other tools to generate a token

**Comment by Vulpesdk** [comment]
I would recommend to do OAuth 2.0 on the collection level instead on individual requests.  Then set a header on the collection like;

![Image](https://github.com/user-attachments/assets/def5a083-a4fc-4bd1-ac4f-1875078631dd)

This will also work for GraphQl requests :)

**Comment by anusree-bruno** [comment]
We’re updating the current OAuth2 implementation in Bruno. Could you take a look at the experimental builds available here: https://github.com/usebruno/bruno-experimental-builds/releases and let us know if this resolves the issue?

**Comment by anusree-bruno** [comment]
Hey @sylvainlegault, could you help me out with a few things?

How is the token being sent in the request — is it through the header, URL, or some other method?
Were you able to make the call successfully by adding the token manually?
If possible, could you share some screenshots of this? Feel free to blur out any sensitive information.

Thanks!

**Comment by VladBozhenok** [comment]
> I would recommend to do OAuth 2.0 on the collection level instead on individual requests. Then set a header on the collection like;
> 
> ![Image](https://github.com/user-attachments/assets/def5a083-a4fc-4bd1-ac4f-1875078631dd)
> 
> This will also work for GraphQl requests :)

It works as a charm, thanks for the help.

Fot those who don't know where to get `access_token_set_by_collection_script` variable, please check this page https://docs.usebruno.com/auth/oauth2/authorization-code

**Comment by anusree-bruno** [comment]
Hey @sylvainlegault, since the main issue has been resolved and we haven’t heard back, we’ll go ahead and close this for now.

If you run into any related issues or have more details to share, feel free to reopen or create a new ticket.

Thanks!

**Comment by syl20lego** [comment]
Thanks @anusree-bruno , didn't have the time to return to this. Will try the next version later and see how it goes.

Thank you!
